### PR TITLE
do not evaluate constants in asm.js, fix IIFE mode detection

### LINKED ISSUE #10283: ConstPlugin invalidates asm.js code
# Bug report

**What is the current behavior?**

ConstPlugin can change asm.js ("use asm") functions and invalidate them by inserting "false", "true", "undefined" values. In asm.js, using these identifiers are not allowed.

**If the current behavior is a bug, please provide the steps to reproduce.**

I'm currently trying to use https://www.npmjs.com/package/viz.js and compiling that with webpack causes it to no longer be valid asm.js code.
But this can be seen with much smaller codes as well, basically with any const conditional. (`0==0` becomes `true`, `truthy() ? 0 : 1` can become `truthy() ? 0 : undefined`, etc)

**What is the expected behavior?**

I think the best would be to simply disable ConstPlugin inside asm.js functions as it would be much harder to properly inject valid asm.js code there. The Parser's scope could be extended to check for "use asm" expressions just like it checks "use strict", and then ConstPlugin could use that info to skip those parts.

**Other relevant information:**
webpack version: 4.41.5

### PR DESCRIPTION
**Pull request by ngg**
fixes #10283

**What kind of change does this PR introduce?**
This PR basically disables the ConstPlugin in asm.js functions.

**Did you add tests for your changes?**
Yes, but it might not be the best place for it, please let me know if there is a better place.

**Does this PR introduce a breaking change?**
It should not cause any compatibility problems.

**What needs to be documented once your changes are merged?**
Probably nothing.

### CODE REVIEW

**Comment by ngg** [comment]
If I open another PR to backport this to webpack 4, would you accept it?

**Comment by ngg** [comment]
There is something strange with this PR, it seems to fix the easy test cases, but on huge emscripten-generated asm.js files it somehow doesn't always recognize the asm.js parts. I'll try to reproduce the issue with a small test case.

**Comment by ngg** [comment]
The following is not recognized as asm.js:
```
(function() { "use asm"; return 0 == 0 ? 1 : 0; })();
```

This is true for "use strict" as well, it is not recognized properly in these cases, I'll try to understand why.

File: lib/javascript/JavascriptParser.js [hunk 1]
```diff
@@ -2172,6 +2173,7 @@ class JavascriptParser extends Parser {
 				this.setVariable(params[i].name, varInfo);
 			}
 			if (functionExpression.body.type === "BlockStatement") {
+				this.detectMode(functionExpression.body.body);
```
Comment (ngg) [inline lib/javascript/JavascriptParser.js]: for IIFEs the `detectStrictMode` function was missing previously

**Comment by sokra** [comment]
Thanks

**Pull request merged**